### PR TITLE
Allow open types in nameof

### PR DESCRIPTION
**Pull request by alrz**
Proposal: https://github.com/dotnet/csharplang/issues/702

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -1227,13 +1227,20 @@ private BoundExpression BindDefaultExpression(DefaultExpressionSyntax node, Diag
...
             bool hasTypeArguments = arity > 0;
 
-            SeparatedSyntaxList<TypeSyntax> typeArgumentList = node.Kind() == SyntaxKind.GenericName
-                ? ((GenericNameSyntax)node).TypeArgumentList.Arguments
-                : default(SeparatedSyntaxList<TypeSyntax>);
+            GenericNameSyntax genericName;
+            SeparatedSyntaxList<TypeSyntax> typeArgumentList = default(SeparatedSyntaxList<TypeSyntax>);
+            bool isUnboundTypeInNameof = false;
+
+            if (node.Kind() == SyntaxKind.GenericName)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Test failure is due to cascade diagnostic on malformed type argument lists. We probably need to return early when `node.ContainsDiagnostic` is true.

**Comment by jcouv** [comment]
@alrz Let's get input from LDM on the feature. I'll let you know when I hear back. Thanks

**Comment by gafter** [comment]
The parser doesn't know whether `nameof` will bind to a method or bind to the built-in `nameof` operator. How does semantic analysis in this PR handle a method invocation that contains an open type somewhere inside it? I would expect this to require heavy testing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 2]
```diff
@@ -1326,5 +1321,256 @@ public static void Main(string[] args)
...
+        _ = nameof(C<>);
+    }
+}
+";
+            var options = TestOptions.ReleaseDll.WithWarningLevel(0);
+            var comp = CreateStandardCompilation(source, options: options);
+            comp.VerifyDiagnostics(
+                // (7,20): error CS0305: Using the generic type 'C<T>' requires 1 type arguments
+                //         _ = nameof(C<>);
+                Diagnostic(ErrorCode.ERR_BadArity, "C<>").WithArguments("C<T>", "type", "1").WithLocation(7, 20),
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Perhaps we don't want to report this cascading diagnostic?

**Comment by cston** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by alrz** [comment]
Since this feature is not championed yet, I'll close this untill LDM confirm.

**Comment by ulln** [comment]
Is this rejected?

If so, i think at least a short explanation should be given why the language designers did decide so.
And if possible, rejected things should be labeled "Rejected".

**Comment by alrz** [comment]
@ulln No, this is not explicitly rejected, but it is pending LDM decision. See https://github.com/dotnet/csharplang/issues/702

**Comment by alrz** [comment]
However I see @jcouv moved this to ["rejected"](https://github.com/dotnet/roslyn/projects/27#card-4083796) column, so I'm not sure what is the state of the proposal.

**Comment by jcouv** [comment]
Yes, this was discussed in LDM and wasn't championed for 7.x or 8.0. Upcoming LDM notes for Nov 6th should have more details (Mads had conferences and is behind on publishing them). Thanks for your patience.

**Comment by jcouv** [comment]
Notes from LDM as FYI https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-11-06.md

**Comment by alrz** [comment]
>@gafter: Does the PR include proposed name lookup rule specification changes to justify the behavior?

@gafter @jcouv If LDT is interested to champion this, I can open a PR against csharplang for a proposal.

**Pull request closed without merging**